### ISSUE TITLE
atlas: multithreaded BLAS is used despite setting site.cfg to use serial one

### ISSUE THREAD

**Issue by davydden** [numpy/numpy#8294]
with site.cfg
```
[ALL]
rpath=/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib
[atlas]
atlas_libs=satlas
library_dirs=/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib
```
numpy decides for itself to use multithreaded version despite specifically being told to use serial one:
```
atlas_3_10_blas_threads_info:
    libraries = ['satlas', 'tatlas']
    library_dirs = ['/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib']
    define_macros = [('HAVE_CBLAS', None), ('ATLAS_INFO', '"\\"3.10.2\\""')]
    language = c
lapack_opt_info:
    libraries = ['tatlas', 'satlas', 'tatlas']
    library_dirs = ['/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib']
    define_macros = [('ATLAS_INFO', '"\\"3.10.2\\""')]
    language = f77
blas_opt_info:
    libraries = ['satlas', 'tatlas']
    library_dirs = ['/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib']
    define_macros = [('HAVE_CBLAS', None), ('ATLAS_INFO', '"\\"3.10.2\\""')]
    language = c
openblas_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
atlas_3_10_threads_info:
    libraries = ['tatlas', 'satlas', 'tatlas']
    library_dirs = ['/home/davydden/spack/opt/spack/linux-ubuntu16-x86_64/gcc-5.4.0/atlas-3.10.2-5jtibobvt7nlwd3zyviionsn7v4z2npd/lib']
    define_macros = [('ATLAS_INFO', '"\\"3.10.2\\""')]
    language = f77
lapack_mkl_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
```

**Comment by rgommers** [comment]
Yep, that's yet another thing that is hardcoded in ``distutils/system_info.py``:

         atlas_info = get_info('atlas_3_10_threads')
         if not atlas_info:
             atlas_info = get_info('atlas_3_10')
         if not atlas_info:
             atlas_info = get_info('atlas_threads')
         if not atlas_info:
             atlas_info = get_info('atlas')

There's a few more issues about ``site.cfg`` not being respected. There'll have to be a significant overhaul of ``system_info`` at some point. For the moment I have nothing better to offer than the advice to just modify the offending piece of code above.

**Comment by rgommers** [comment]
Searching issues with ``is:issue is:open site.cfg`` gives 15 hits right now .....

**Comment by davydden** [comment]
i see, thanks for the reply. Hopefully it will be fixed at some point.

**Comment by davydden** [comment]
> For the moment I have nothing better to offer than the advice to just modify the offending piece of code above.

frankly, i don't see what would i need to do to turn off any auto-detection, which is, as far as i understand, what's happening in those `get_info()`. 

> There'll have to be a significant overhaul of system_info at some point. 

i think it would be better if there is an opt-out from all auto-detection related to Blas/Lapack. If a user provides a list of `mkl` libs -- take it; If it is a list of libraries for `atlas` -- take it, etc. If a user provides `lib` folder only, then auto-detection should kick in.

**Comment by rgommers** [comment]
> frankly, i don't see what would i need to do to turn off any auto-detection

**Comment by rgommers** [comment]
> frankly, i don't see what would i need to do to turn off any auto-detection

Indeed, that's not easy. My suggestion only works if you know you want only 1 fixed set of libs to link against. In your case, if you want the non-threaded ``satlas`` to be used, this should do the job:

    diff --git a/numpy/distutils/system_info.py b/numpy/distutils/system_info.py
    index b8520ac..05af2e1 100644
    --- a/numpy/distutils/system_info.py
    +++ b/numpy/distutils/system_info.py
    @@ -1480,9 +1480,7 @@ class lapack_opt_info(system_info):
                 self.set_info(**openblas_info)
                 return
     
    -        atlas_info = get_info('atlas_3_10_threads')
    -        if not atlas_info:
    -            atlas_info = get_info('atlas_3_10')
    +        atlas_info = get_info('atlas_3_10')
             if not atlas_info:
                 atlas_info = get_info('atlas_threads')
             if not atlas_info:

> i think it would be better if there is an opt-out from all auto-detection related to Blas/Lapack.

I agree, that would be useful. That does require quite a bit of surgery though, since the sections in ``site.cfg`` are parsed and stored, then accessed from *within* the code that now does the auto-detection.

**Comment by davydden** [comment]
>  In your case, if you want the non-threaded satlas to be used

i see what this is doing, but it won't help me because it has to work based on `site.cfg`. Look at this PR https://github.com/LLNL/spack/pull/2365 in SupercomputerPACKage manager (Spack) to get an idea why. The choice of flavour of Blas/Lapack (Atlas/MKL/Openblas or their variants like threaded or not, 32bit or 64) is up to a user. In case of Atlas, the resulting set of libraries is written to `atlas_libs=`. So the ultimate patch needs to make `numpy` respect libraries provided within this field. But as you said above it's not straight forward and requires quite a lot of changes in the build system. So i guess for now one has to live with these inconsistencies and hope that it will get fixed in next releases.

**Comment by rgommers** [comment]
Thanks @davydden, clear that you need something without changes to numpy itself.

**Comment by rgommers** [comment]
After going through the whole set of ``site.cfg`` issues, I'm almost tempted to leave that as is for backwards compat and create a new ``.cfg`` config file with sane behavior. There are so many corner cases and people with workarounds for current behavior that it'll be very hard to fix otherwise.

**Comment by davydden** [comment]
> ... create a new .cfg config file with sane behavior.

are there any updates on this front?

**Comment by mattip** [comment]
Closing, site.cfg and numpy.distutils have bene removed
